KAFKA-16001: Migrated ConsumerNetworkThreadTest away from ConsumerTestBuilder

brenden20 opened the pull request:
Completely migrated ConsumerNetworkThreadTest away from ConsumerTestBuilder and removed all usages of spy objects and replaced with mocks. Removed testEnsureMetadataUpdateOnPoll() since it was doing integration testing. Also I added 2 new tests to get more complete test coverage of ConsumerNetworkThread. All tests are passing

brenden20 committed:
Migrated ConsumerTestBuilder

Removed dependencies upon ConsumerTestBuilder

brenden20 committed:
Removed changes to OffsetFetchRequestState

Removed change of toString to toStringBase, moving change to separate branch

brenden20 committed:
Added test for RMs

Added test for RMs, seeing if pollWaitTimeMs is computed correctly

brenden20 committed:
Cleaning up comments

Cleaning up comments

brenden20 committed:
Added test for wait time computation

Added a new param test to ensure ConsumerNetworkThread wait times are being computed correctly.

brenden20 committed:
Test removal

Removed all redundant/integration tests

brenden20 committed:
Changed variables names

Changed variables names in testConsumerNetworkThreadWaitTimeComputations to be more descriptive

brenden20 committed:
Removed vars and methods with no usage

Removed vars and methods with no usage

brenden20 committed:
Removed unused imports

Removed unused imports

brenden20 committed:
Updating comments

Updating comments

brenden20 committed:
Comments

Comments

brenden20 committed:
Merge branch 'apache:trunk' into 16001

brenden20 committed:
Merge branch 'apache:trunk' into 16001

brenden20 committed:
Updated constructor

Updated constructor to include correct parameters

brenden20 committed:
Added tests back

Added previously removed tests and changed some imports and edited the constructor

brenden20 committed:
Minor updates to tests

Did some minor updates, changed testMetadataUpdateEvent since it was integration testing

brenden20 committed:
Constructor and test updates

Updated constructor, removed unnecessary variables. Also updated a few tests to make them work or to make them unit tests rather than integration tests

brenden20 committed:
Updated comment

Updated comment on testEnsureMetadataUpdateOnPoll

brenden20 committed:
Merge branch '16001' of https://github.com/brenden20/kafka into 16001

brenden20 committed:
Remove comment

Remove comment

brenden20 committed:
Added testEnsureCloseStopsRunningThread(), updated others

Added testEnsureCloseStopsRunningThread.

Updated testConsumerNetworkThreadWaitTimeComputations to also test that networkClientDelegate polls with correct times.

Removed one instance of a spy object, replaced with mock.

brenden20 committed:
Cleaning up warnings

Cleaning up warnings, changed lambda expressions to explicit method references

brenden20 committed:
Updated testAssignmentChangeEvent()

Removed tests from this test that seem to be more like integration testing

brenden20 committed:
Removed testEnsureMetadataUpdateOnPoll()

Removed testEnsureMetadataUpdateOnPoll().

This is an integration test, tried many different ways to still get it to pass but could not. The issue is that for the metadata. updateWithCurrentRequestVersion() method to be invoked, many objects must not be mocks. This causes issues when trying to use verify(), as it can only take a mock, not an actual instantiated object.

I also cleaned up the code a little bit

brenden20 committed:
Changed MockClient

Changed instantiation of MockClient to not include metadata in constructor

brenden20 committed:
Moved all changes into original file

Moved all code from ConsumerNetworkThreadUnitTest to ConsumerNetworkThreadTest and removed ConsumerNetworkThreadUnitTest

brenden20 committed:
Comment removal

Comment removal

kirktrue commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkThreadTest.java:
```diff
@@ -149,20 +157,28 @@ public void testStartupAndTearDown() throws InterruptedException {
                 "The consumer network thread did not stop within " + DEFAULT_MAX_WAIT_MS + " ms");
     }
 
+    @Test
+    void testRequestManagersArePolledOnce() {
+        consumerNetworkThread.runOnce();
+        requestManagers.entries().forEach(rmo -> rmo.ifPresent(rm -> verify(rm, times(1)).poll(anyLong())));
+        requestManagers.entries().forEach(rmo -> rmo.ifPresent(rm -> verify(rm, times(1)).maximumTimeToWait(anyLong())));
+        verify(networkClientDelegate).poll(anyLong(), anyLong());
+    }
+
     @Test
     public void testApplicationEvent() {
         ApplicationEvent e = new PollEvent(100);
         applicationEventsQueue.add(e);
         consumerNetworkThread.runOnce();
-        verify(applicationEventProcessor, times(1)).process(e);
+        verify(applicationEventProcessor).process(e);
```
I'm curious as to this change. Is this something that was incorrect as is? 🤔

brenden20 replied:
I checked Mockito documentation and adding times(1) is redundant, so not really a big deal either way to keep it or remove it. Do let me know though if there is a stylistic preference.

kirktrue commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/CommitRequestManager.java:
```diff
@@ -1290,4 +1290,4 @@ static class MemberInfo {
         }
     }
-}
+}
```
We should revert/fix this change as it's whitespace only.

kirktrue commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkThreadTest.java:
```diff
@@ -53,89 +39,111 @@
+    @ParameterizedTest
+    @ValueSource(longs = {1, 100, 1000, 4999, 5001})
+    public void testConsumerNetworkThreadWaitTimeComputations(long exampleTime) {
+        List<Optional<? extends RequestManager>> requestManagersList = new ArrayList<>();
+        requestManagersList.add(Optional.of(coordinatorRequestManager));
+        when(requestManagers.entries()).thenReturn(requestManagersList);
```
Nit, but could this be:

```suggestion
        when(requestManagers.entries()).thenReturn(Collections.singletonList(Optional.of(coordinatorRequestManager)));
```

kirktrue commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkThreadTest.java:
```diff
@@ -268,32 +281,31 @@ void testPollResultTimer() {
         // Initial value before runOnce has been called
         assertEquals(ConsumerNetworkThread.MAX_POLL_TIMEOUT_MS, consumerNetworkThread.maximumTimeToWait());
+
+        when(requestManagers.entries()).thenReturn(list);
```
Minor, but it seems like this line could be more succinct like:

```suggestion
        when(requestManagers.entries()).thenReturn(Collections.singletonList(Optional.of(heartbeatRequestManager)));
```

kirktrue commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkThreadTest.java:
```diff
@@ -379,13 +403,4 @@ private MockClient.RequestMatcher offsetCommitRequestMatcher(final Map<TopicPart
-        return topicPartitionOffsets;
-    }
-}
+}
```
```suggestion
}

```

kirktrue commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkThreadTest.java:
```diff
@@ -314,7 +326,7 @@ void testEnsureEventsAreCompleted() {
         coordinatorRequestManager.markCoordinatorUnknown("test", time.milliseconds());
         client.prepareResponse(FindCoordinatorResponse.prepareResponse(Errors.NONE, "group-id", node));
         prepareOffsetCommitRequest(new HashMap<>(), Errors.NONE, false);
-        CompletableApplicationEvent<Void> event1 = spy(new AsyncCommitEvent(Collections.emptyMap()));
+        CompletableApplicationEvent<Void> event1 = mock(AsyncCommitEvent.class);
```
👍

kirktrue commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkThreadTest.java:
```diff
@@ -327,8 +339,11 @@ void testEnsureEventsAreCompleted() {
         assertTrue(applicationEventsQueue.isEmpty());
     }
 
+    // Look into this one
```
Can you elaborate on this comment?

brenden20 replied:
My mistake leaving that there, that was a comment for myself that I forgot to remove

kirktrue reviewed (changes requested):
Thanks @brenden20. This is a welcome change and I only have minor nits at this point.

kirktrue commented:
Looks like there are some checkstyle failures due to the use of wildcard imports.

kirktrue commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkThreadTest.java:
```diff
@@ -53,89 +39,111 @@
-import static org.junit.jupiter.api.Assertions.assertFalse;
-import static org.junit.jupiter.api.Assertions.assertTrue;
-import static org.mockito.ArgumentMatchers.any;
+import static org.junit.jupiter.api.Assertions.*;
```
```suggestion
import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.mockito.ArgumentMatchers.any;
```

kirktrue commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkThreadTest.java:
```diff
@@ -53,89 +39,111 @@
-import static org.mockito.Mockito.times;
-import static org.mockito.Mockito.verify;
-import static org.mockito.Mockito.when;
+import static org.mockito.Mockito.*;
```
Just bring these explicit imports back to make checkstyle stop complaining

```suggestion
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.doAnswer;
import static org.mockito.Mockito.doThrow;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;
```

kirktrue reviewed (changes requested):
Left some suggestions as to the checkstyle issues.

You should be able to run this command locally to verify the checkstyle rules pass:

```
./gradlew check -x test
```

brenden20 committed:
Cleaned up testConsumerNetworkThreadWaitTimeComputations

Made the first when() statement cleaner

brenden20 committed:
Cleaned up testMaximumTimeToWait()

Condensed a few lines of code into one cleaner line

brenden20 committed:
Cleaned imports and comment removal

Cleaned up wildcard imports can removed extraneous comment

brenden20 committed:
Fix whitespace

Co-authored-by: Kirk True <kirk@kirktrue.pro>

brenden20 committed:
Merge branch 'apache:trunk' into 16001

philipnee commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkThreadTest.java:
```diff
@@ -16,33 +16,19 @@
-import org.apache.kafka.common.requests.OffsetCommitRequest;
-import org.apache.kafka.common.requests.OffsetCommitResponse;
-import org.apache.kafka.common.requests.RequestTestUtils;
+import org.apache.kafka.common.requests.*;
```
in kafka we don't use wildcard import, could you adjust the setting in the IDE? I think we also want to avoid rearranging imports since these changes aren't necessary relevant to this PR.

philipnee commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkThreadTest.java:
```diff
@@ -53,19 +39,11 @@
-import java.util.Iterator;
-import java.util.Map;
-import java.util.Optional;
+import java.util.*;
```
same, we don't use wild card.

philipnee commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkThreadTest.java:
```diff
@@ -74,68 +52,105 @@
+        this.offsetsRequestManager = mock(OffsetsRequestManager.class);
+        this.heartbeatRequestManager = mock(HeartbeatRequestManager.class);
+        this.coordinatorRequestManager = mock(CoordinatorRequestManager.class);
+        this.applicationEventsQueue = new LinkedBlockingQueue<>();
```
can we mock this as well?

philipnee commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkThreadTest.java:
```diff
@@ -74,68 +52,105 @@
+        this.applicationEventReaper = mock(CompletableEventReaper.class);
+        this.client = new MockClient(time);
+
+        this.networkClient = new NetworkClientDelegate(
```
i think i might have asked you to use MockClient here.  since we aren't necessary testing request sending etc., can we just mock the networkClientDelegate?

lianetm replied:
this is a good point, what was the resolution? I expect we should be able to mock everything except the ConsumerNetworkThread (but I do see how we have tests asserting on the network client so it would require untangling that, is that the reason why we're still not mocking the NetworkClientDelegate?). OK for me to leave it as it is and do a follow-up btw, just want to understand.

philipnee commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkThreadTest.java:
```diff
@@ -74,68 +52,105 @@
+
+    @ParameterizedTest
+    @ValueSource(longs = {1, 100, 1000, 4999, 5001})
+    public void testConsumerNetworkThreadWaitTimeComputations(long exampleTime) {
```
can we test it with 2 request managers to ensure Math.min is applied?

philipnee commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkThreadTest.java:
```diff
@@ -149,20 +164,28 @@ public void testStartupAndTearDown() throws InterruptedException {
+    @Test
+    void testRequestManagersArePolledOnce() {
+        consumerNetworkThread.runOnce();
+        requestManagers.entries().forEach(rmo -> rmo.ifPresent(rm -> verify(rm, times(1)).poll(anyLong())));
```
i believe times(1) is the default, could you verify this?

brenden20 replied:
It is, removing instances of times(1)

philipnee commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkThreadTest.java:
```diff
@@ -149,20 +164,28 @@ public void testStartupAndTearDown() throws InterruptedException {
     }
 
     @Test
     public void testMetadataUpdateEvent() {
```
i think can probably remove this teset as it is doing exactly the same thing as `testApplicationEvent`.  We should probably try to find a similar test to ensure the code path is covered.

brenden20 replied:
Removed

philipnee commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkThreadTest.java:
```diff
@@ -207,7 +230,7 @@ public void testResetPositionsProcessFailureIsIgnored() {
         ResetPositionsEvent event = new ResetPositionsEvent(calculateDeadlineMs(time, 100));
         applicationEventsQueue.add(event);
-        assertDoesNotThrow(() -> consumerNetworkThread.runOnce());
+        assertDoesNotThrow(consumerNetworkThread::runOnce);
```
can we revert the irrelevant change? also I think we can remove this test and see if this is covered elsewhere.

philipnee commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkThreadTest.java:
```diff
@@ -231,10 +254,7 @@ public void testAssignmentChangeEvent() {
-        verify(commitRequestManager, times(1)).updateAutoCommitTimer(currentTimeMs);
-        // Assignment change should generate an async commit (not retried).
-        verify(commitRequestManager, times(1)).maybeAutoCommitAsync();
+        verify(networkClientDelegate, times(1)).poll(anyLong(), anyLong());
```
can we be consistent at using times(1) here? above you removed times(1) but it seems to be inconsistently applied elsewhere.

philipnee commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkThreadTest.java:
```diff
@@ -339,6 +358,15 @@ void testRunOnceInvokesReaper() {
         verify(applicationEventReaper).reap(any(Long.class));
     }
 
+    private HashMap<TopicPartition, OffsetAndMetadata> mockTopicPartitionOffset() {
```
try to use Map instead of HashMap

philipnee commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkThreadTest.java:
```diff
@@ -339,6 +358,15 @@ void testRunOnceInvokesReaper() {
+    private HashMap<TopicPartition, OffsetAndMetadata> mockTopicPartitionOffset() {
+        final TopicPartition t0 = new TopicPartition("t0", 2);
+        final TopicPartition t1 = new TopicPartition("t0", 3);
+        HashMap<TopicPartition, OffsetAndMetadata> topicPartitionOffsets = new HashMap<>();
```
1. this can be final as well
2. let's use Map on the left side.

philipnee reviewed (changes requested):
hi @brenden20 - left some comments. lmk what do you think.

brenden20 committed:
Implementing suggestions from PR

Changed some variable stuff, undid unnecessary changes, removed redundant code, fixed imports, and fixed checktyle violation

brenden20 committed:
Add comment

brenden20 committed:
Todo comment

brenden20 committed:
Updated testConsumerNetworkThreadWaitTimeComputations()

Updated testConsumerNetworkThreadWaitTimeComputations() to include a second request manager to ensure correct computation

brenden20 committed:
Added new test

Added testEnsureSendUnsentRequestPollWIthZeroRunsOnce(), it checks that when ConsumerNetworkThread.sendUnsentRequests() with a timer with timeoutMs = 0 will only poll the networkClientDelegate once

brenden20 committed:
Updated testConsumerNetworkThreadWaitTimeComputations()

Updated testConsumerNetworkThreadWaitTimeComputations(), removed some unnecessary test cases

brenden20 committed:
Merge branch 'apache:trunk' into 16001

brenden20 committed:
Revert "Merge branch 'apache:trunk' into 16001"

This reverts commit cf1b5e78096899281b4b523b8dc95eb042f267d8, reversing
changes made to 4191c2d2cccdadd647ee950c81e201029ab08715.

lianetm commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkThreadTest.java:
```diff
@@ -150,19 +209,19 @@ public void testStartupAndTearDown() throws InterruptedException {
+    void testRequestManagersArePolledOnce() {
         consumerNetworkThread.runOnce();
-        verify(applicationEventProcessor, times(1)).process(e);
+        requestManagers.entries().forEach(rmo -> rmo.ifPresent(rm -> verify(rm).poll(anyLong())));
```
I think we're missing the expectations for the requestManagers.entries() to make sure there are some managers and we actually verify something, so we need  `when(requestManagers.entries()).thenReturn(..` with some managers.

lianetm commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkThreadTest.java:
```diff
@@ -150,19 +209,19 @@ public void testStartupAndTearDown() throws InterruptedException {
-        verify(applicationEventProcessor, times(1)).process(e);
+        requestManagers.entries().forEach(rmo -> rmo.ifPresent(rm -> verify(rm).poll(anyLong())));
+        requestManagers.entries().forEach(rmo -> rmo.ifPresent(rm -> verify(rm).maximumTimeToWait(anyLong())));
+        verify(networkClientDelegate).poll(anyLong(), anyLong());
```
This test seems to leave out one of the core actions of the `runOnce`:

1. poll managers (this generates the requests) -> covered by the test
2. add newly generated requests are added to the network client -> not covered
3. poll network client to send the requests -> covered by the test

Step 3 wouldn't have the effect we want (send the request) if step 2 does not happen, so what about adding a `verify(networkClientDelegate).addAll(..` before verifying the network client poll, to make sure we cover the whole flow of how requests move from the managers to the network layer on run once?

brenden20 replied:
Thank you for the feedback, I have implemented this suggestion as well as the previous one for this test

lianetm commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkThreadTest.java:
```diff
@@ -150,19 +209,19 @@ public void testStartupAndTearDown() throws InterruptedException {
     @Test
-    public void testMetadataUpdateEvent() {
-        ApplicationEvent e = new NewTopicsMetadataUpdateRequestEvent();
+    public void testApplicationEvent() {
```
Not introduced by this PR, but reviewing this one I noticed that we have lots of similar tests for checking that runOnce processes the events (ex. testResetPositionsEventIsProcessed, testSyncCommitEvent, testAsyncCommitEvent...), all doing the same (add event, run once, check is processed). Could we parametrize this and have a single test maybe? (same shape as this one but receiving the event as param). Should allow us to remove lots that, in the end this component does not care about the specific events, it's just responsible for processing whatever event is added.

philipnee replied:
+1

brenden20 replied:
I tried this out, it seems like parameterizing the tests into one will not work, it keeps giving me an error saying that the parameters must be constant. I did some research, it looks like it can be done, but would require an enum class to get it to work. Not sure if this is something we would want to do just for these few tests, but let me know.

lianetm replied:
I had something like this in mind:
```
    @ParameterizedTest
    @MethodSource("appEvents")
    public void testEventIsProcessed(ApplicationEvent e) {
        applicationEventsQueue.add(e);
        consumerNetworkThread.runOnce();
        verify(applicationEventProcessor).process(any(e.getClass()));
        assertTrue(applicationEventsQueue.isEmpty());
    }

    private static Stream<Arguments> appEvents() {
        return Stream.of(
            Arguments.of(new PollEvent(100)),
            Arguments.of(new NewTopicsMetadataUpdateRequestEvent()));
    }
```
(extending `appEvents` with all the event instances that are individually created in each testxxxIsProcessed so we can replace them all with this single parametrized test)

lianetm commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkThreadTest.java:
```diff
@@ -329,6 +382,8 @@ void testEnsureEventsAreCompleted() {
     @Test
     void testCleanupInvokesReaper() {
+        LinkedList<NetworkClientDelegate.UnsentRequest> queue = new LinkedList<>();
+        when(networkClientDelegate.unsentRequests()).thenReturn(queue);
```
totally needed, but heads-up, there is another [PR](https://github.com/apache/kafka/pull/16156) in-flight where we're trying to change the check behind this, so we'll need to update this expectation here if that goes in first.

brenden20 replied:
Thanks for the heads up, I will keep that in mind

lianetm replied:
it just got merged, so worth rebasing and updating the test as needed ;)

brenden20 replied:
I will get that done, thank you for letting me know!

lianetm commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkThread.java:
```diff
@@ -289,10 +289,11 @@ private void closeInternal(final Duration timeout) {
         }
     }
 
+    // Add test to see if poll() is run once with timer of 0
```
is this something you intend to do on this PR or separately? Either way, I would suggest we remove the comment and address it in this PR or have a jira to follow-up separately if needed please.

brenden20 replied:
I forgot to remove that, removed now

lianetm commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkThreadTest.java:
```diff
@@ -270,30 +326,27 @@ void testPollResultTimer() {
-        consumerNetworkThread.runOnce();
-        verify(metadata, times(1)).updateWithCurrentRequestVersion(eq(metadataResponse), eq(false), anyLong());
-    }
+        List<Node> list = new ArrayList<>();
+        list.add(new Node(0, "host", 0));
+        when(cluster.nodes()).thenReturn(list);
```
what about simplifying to a single `when(cluster.nodes()).thenReturn(Collections.singletonList(new Node(0, "host", 0)));` ?

lianetm commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkThreadTest.java:
```diff
@@ -270,30 +326,27 @@ void testPollResultTimer() {
-        testBuilder.requestManagers.entries().forEach(rmo -> rmo.ifPresent(rm -> verify(rm, times(1)).maximumTimeToWait(anyLong())));
-        verify(networkClient, times(1)).poll(anyLong(), anyLong());
-    }
+    void testEnsureEventsAreCompleted() {
```
Not introduced by this PR but since we're improving here, I find that this test does not bring any value, because it's testing something that it's not the responsibility of the `ConsumerNetworkThread`, so we end up testing something we're mocking ourselves with the `doAnswer`. What do you think?

I had suggested to remove it completely. See my original comment about it on this other [PR](https://github.com/apache/kafka/pull/15640/files#r1608897877). If you agree I would say we remove it instead of keep having to maintain it.

brenden20 replied:
I went ahead and removed this test

lianetm commented:
Hey @brenden20, very nice improvement! Left a few comments. Thanks!

brenden20 committed:
Update ConsumerNetworkThread.java

brenden20 committed:
testRequestManagersArePolledOnce()Updated

brenden20 committed:
Small test updates

brenden20 committed:
Reapply "Merge branch 'apache:trunk' into 16001"

This reverts commit 5fb7304892ecf5662a23b5c65111c79b012007ab.

brenden20 committed:
Add comment

brenden20 commented:
@lianetm @kirktrue @philipnee, I have implemented all suggestions except for test removals since I think if we do that, it should go on another ticket as to not bog this ticket down. All tests are passing as well, so let me know if there is anything else I should take care of or if this PR is ready for review. Thank you all for the reviews!

philipnee commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkThreadTest.java:
```diff
@@ -78,64 +81,120 @@
+        this.metadata = mock(ConsumerMetadata.class);
+        this.applicationEventProcessor = mock(ApplicationEventProcessor.class);
+        this.applicationEventReaper = mock(CompletableEventReaper.class);
+        this.client = mock(MockClient.class);
```
new MockClient() ?

philipnee commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkThreadTest.java:
```diff
@@ -78,64 +81,120 @@
+    }
+
+    @Test
+    public void testEnsureSendUnsentRequestPollWIthZeroRunsOnce() {
```
sorry didn't catch this: `sendUnsentRequests` is only invoked in cleanup(), so we probably should test cleanup(), not `sendUnsentRequests`.

Here's an explanation on the internet: 
```
You generally don't unit test private methods directly. Since they are private, consider them an implementation detail. Nobody is ever going to call one of them and expect it to work a particular way.
```

brenden20 replied:
Looking into this, I am not sure there is a way I can test this functionality without making changes to the methods in ```ConsumerNetworkThread```. To poll with a time of 0, the timer being used must be instantiated with a 0 as ```timeoutMs```. However, the timer that ```ConsumerNetworkThread``` uses is a local variable ```closeTimeout```, which is private and is only modified by a single private method that is unrelated to ```cleanup()```. 

Let me know what you think

lianetm replied:
I agree with @philipnee 's point, but given that it would require changes in the ConsumerNetworkThread and the test is not introduced by this PR I would lean towards doing it in a follow-up PR (and keep the scope of this PR limited to the test changes only)

philipnee replied:
+1

philipnee commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkThreadTest.java:
```diff
@@ -78,64 +81,120 @@
+
+    @ParameterizedTest
+    @ValueSource(longs = {100, 4999, 5001})
+    public void testConsumerNetworkThreadWaitTimeComputations(long exampleTime) {
```
nit: PollTime instead of WaitTime

philipnee commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkThreadTest.java:
```diff
@@ -150,19 +209,27 @@ public void testStartupAndTearDown() throws InterruptedException {
-    public void testApplicationEvent() {
-        ApplicationEvent e = new PollEvent(100);
-        applicationEventsQueue.add(e);
+    void testRequestManagersArePolledOnce() {
```
let's just make this public void for consistency.

philipnee commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkThread.java:
```diff
@@ -292,7 +292,8 @@ private void closeInternal(final Duration timeout) {
      */
-    private void sendUnsentRequests(final Timer timer) {
+    // Visible for testing
+    protected void sendUnsentRequests(final Timer timer) {
```
i was wondering if we could just test cleanup()

brenden20 committed:
Implementing suggestions

brenden20 committed:
Updated test name

brenden20 committed:
Condensed multiple tests into 1

Condensed multiple tests into 1 that is parameterized and performs the same functionality

brenden20 committed:
Fixing conflict

brenden20 committed:
Revert "Fixing conflict"

This reverts commit d443311ec4c980b24eaaad9a5ba80df64fb7de85.

brenden20 committed:
Merge branch 'trunk' into 16001

brenden20 committed:
Updated new test to work with current file

lianetm commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkThreadTest.java:
```diff
@@ -74,68 +79,113 @@
+
+    @Test
+    public void testEnsureCloseStopsRunningThread() {
+        // consumerNetworkThread.running is set to true in its constructor
+        assertTrue(consumerNetworkThread.isRunning());
```
```suggestion
        assertTrue(consumerNetworkThread.isRunning(), "ConsumerNetworkThread should start running when created");
```
or similar msg, but it seems cleaner and more helpful in case of failure

philipnee replied:
+1

lianetm commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkThreadTest.java:
```diff
@@ -74,68 +79,113 @@
+        // consumerNetworkThread.running is set to true in its constructor
+        assertTrue(consumerNetworkThread.isRunning());
+
+        // close() should make consumerNetworkThread.running false by calling closeInternal(Duration timeout)
```
ditto, I would add this as a high level assertion msg if you think it helps, and remove the comment (I don't think we need that much detail about what we expect, it should simply stop on close)

philipnee replied:
+1

lianetm commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkThreadTest.java:
```diff
@@ -150,35 +200,44 @@ public void testStartupAndTearDown() throws InterruptedException {
-    public void testApplicationEvent() {
-        ApplicationEvent e = new PollEvent(100);
-        applicationEventsQueue.add(e);
+    public void testRequestManagersArePolledOnce() {
```
nit: this name has diverged from the test content I would say. What about `testRequestsTransferFromManagersToClientOnThreadRun` or something like it?

philipnee replied:
i don't quite get the TransferFromManagersToClient part, perhaps you mean by `        verify(networkClientDelegate).addAll(any(NetworkClientDelegate.PollResult.class));` ?

i agree that the test verifies a lot  more calls than just verifying rms are `PolledOnce`.

lianetm replied:
Yes, this part we're testing gets the requests from the managers and adds them to the network client (that's the transfer I was referring to). But whatever name we prefer, the point is to reflect what we're testing (which is not "managersPolledOnce")

lianetm commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkThreadTest.java:
```diff
@@ -353,13 +379,22 @@ void testSendUnsentRequest() {
         assertFalse(networkClient.unsentRequests().isEmpty());
         assertFalse(client.hasInFlightRequests());
-        consumerNetworkThread.cleanup();
+        consumerNetworkThread1.cleanup();
```
nit: up to you but we really don't need this change, and could keep the var name just as `consumerNetworkThread`

lianetm commented:
Hey @brenden20 , thanks for the updates! Left some minor comments.

philipnee commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkThreadTest.java:
```diff
@@ -150,35 +200,44 @@ public void testStartupAndTearDown() throws InterruptedException {
         applicationEventsQueue.add(e);
         consumerNetworkThread.runOnce();
-        verify(applicationEventProcessor).process(any(AsyncCommitEvent.class));
+        verify(applicationEventProcessor).process(any(e.getClass()));
```
this is not right, the method call is `applicationEventProcessor.process();` in runOnce()

philipnee replied:
also need to check this part:
```
  if (event instanceof CompletableEvent)
                    applicationEventReaper.add((CompletableEvent<?>) event);
```

philipnee commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkThreadTest.java:
```diff
@@ -150,35 +200,44 @@ public void testStartupAndTearDown() throws InterruptedException {
-        consumerNetworkThread.runOnce();
-        verify(applicationEventProcessor).process(any(SyncCommitEvent.class));
+    private static Stream<Arguments> appEvents() {
+        Time time1 = new MockTime();
```
can we just call this `time`?

philipnee replied:
you can just this.time ?

philipnee commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkThreadTest.java:
```diff
@@ -341,6 +356,17 @@ void testRunOnceInvokesReaper() {
 
     @Test
     void testSendUnsentRequest() {
+        ConsumerNetworkThread consumerNetworkThread1 = new ConsumerNetworkThread(
```
can you just use `this.consumerNetworkThread` ?

philipnee commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkThreadTest.java:
```diff
@@ -341,6 +356,17 @@ void testRunOnceInvokesReaper() {
+                () -> networkClient,
+                () -> requestManagers
+        );
+        consumerNetworkThread1.initializeResources();
```
you can also remove this line. since it's in the setup()

philipnee commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkThreadTest.java:
```diff
@@ -74,68 +79,113 @@
                 () -> applicationEventProcessor,
-                () -> testBuilder.networkClientDelegate,
-                () -> testBuilder.requestManagers
+                () -> networkClientDelegate,
```
you created two networkClientDelegate instances here - you should pick one to use.

philipnee commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkThreadTest.java:
```diff
@@ -74,68 +79,113 @@
+
+    ConsumerNetworkThreadTest() {
+        ConsumerConfig config = mock(ConsumerConfig.class);
+        this.time = new MockTime();
```
for aesthetic purposes, can we group mock instances and non-mock ones?

brenden20 committed:
Implementing PR suggestions

Removed test that was suggested to be removed, changed some test and method names, and updated a few tests

brenden20 committed:
Updated constructor to look a little better

brenden20 committed:
Updated testApplicationEventIsProcessed()

Updated testApplicationEventIsProcessed() to ensure that for completable events, the applicationEventReaper adds those

brenden20 committed:
Fixed checkstyle violation

lianetm commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkThreadTest.java:
```diff
@@ -150,35 +184,49 @@ public void testStartupAndTearDown() throws InterruptedException {
+        Map<TopicPartition, OffsetAndMetadata> offset = mockTopicPartitionOffset();
+        final long currentTimeMs = time1.milliseconds();
+
+        // use 500 for deadlineMs
```
what's this comment for?

brenden20 replied:
Mistake leaving that there, fixed now

lianetm commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkThreadTest.java:
```diff
@@ -258,77 +272,31 @@ void testPollResultTimer() {
-    }
-
-    @Test
-    void testEnsureEventsAreCompleted() {
```
yay! thanks for removing this! it's been hanging around for too long without value

lianetm commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkThreadTest.java:
```diff
@@ -212,40 +251,15 @@ public void testResetPositionsProcessFailureIsIgnored() {
-    }
-
     @Test
     void testPollResultTimer() {
```
This test is all about the `NetworkClientDelegate`, no interaction at all with the `ConsumerNetworkThread` right? Should be completely moved to `NetorkClientDelegateTest` I would say (this PR or separately, as you wish)

brenden20 replied:
Good point, I will make a separate ticket and PR to handle this

lianetm commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkThreadTest.java:
```diff
@@ -341,6 +309,24 @@ void testRunOnceInvokesReaper() {
 
     @Test
     void testSendUnsentRequest() {
+        NetworkClientDelegate networkClientDelegate = new NetworkClientDelegate(
```
This could all be greatly simplified. What about replacing the whole test with something like:

```
when(networkClientDelegate.hasAnyPendingRequests()).thenReturn(true).thenReturn(true).thenReturn(false);
        consumerNetworkThread.cleanup();
        verify(networkClientDelegate, times(2)).poll(anyLong(), anyLong());
```

The responsibility of the component is only to poll the network client as long as there are pending requests, so with the above we verify that (no need to get into creating an actual network client and assert on it, etc). Makes sense?

brenden20 replied:
Got that done! Thank you

lianetm commented:
I completed another pass @brenden20 , left some comments to consider getting rid of some instances that we couldn't still mock but seems possible. Thanks!

brenden20 committed:
Fixed reminder comment

brenden20 committed:
Update test scope and testSendUnsentRequests()

brenden20 committed:
Cleaned up unused imports and unused methods

Cleaned up unused imports and unused methods,

lianetm commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkThreadTest.java:
```diff
@@ -258,155 +252,66 @@ void testPollResultTimer() {
-        assertFalse(networkClient.unsentRequests().isEmpty());
-        assertFalse(client.hasInFlightRequests());
+    public void testSendUnsentRequests() {
+        when(networkClientDelegate.unsentRequests()).thenReturn(new LinkedList<>());
```
we don't need this (the call is encapsulated in the `hasAnyPendingRequests` that we're setting up right below)

lianetm commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkThreadTest.java:
```diff
@@ -258,155 +252,66 @@ void testPollResultTimer() {
         consumerNetworkThread.cleanup();
 
-        assertTrue(networkClient.unsentRequests().isEmpty());
+        verify(networkClientDelegate, times(2)).poll(anyLong(), anyLong());
+        assertTrue(networkClientDelegate.unsentRequests().isEmpty());
         assertFalse(client.hasInFlightRequests());
```
```suggestion
        verify(networkClientDelegate, times(2)).poll(anyLong(), anyLong());
```

We don't need to assert on the mocked clients. As long as we can verify that the consumer network thread is calling the client poll, the rest is the responsibility of the networkClientDelegate

brenden20 replied:
Makes sense, fixed

lianetm commented:
Hey @brenden20 , thanks for the update, just 2 minor comments left. Could you please get trunk latests changes and solve the conflicts? Thanks!

brenden20 committed:
Merge branch 'trunk' into 16001

brenden20 committed:
Update testSendUnsentRequests()

brenden20 commented:
@lianetm implemented suggestions and merged with trunk for no conflicts

brenden20 committed:
Removed unused import

lianetm commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkThreadTest.java:
```diff
@@ -211,178 +232,80 @@ public void testResetPositionsProcessFailureIsIgnored() {
+        req.setTimer(time, DEFAULT_REQUEST_TIMEOUT_MS);
+
+        // purposely setting a non-MAX time to ensure it is returning Long.MAX_VALUE upon success
+        NetworkClientDelegate.PollResult success = new NetworkClientDelegate.PollResult(
+                10,
+                Collections.singletonList(req));
+        assertEquals(10, networkClientDelegate.addAll(success));
+
+        NetworkClientDelegate.PollResult failure = new NetworkClientDelegate.PollResult(
+                10,
+                new ArrayList<>());
+        assertEquals(10, networkClientDelegate.addAll(failure));
```
this seems like a duplicate, not aligned with the comment, what's the intention?

brenden20 replied:
It sems some of my changes were overwritten when I merged, I am fixing that up

brenden20 committed:
Fixed some merge issues

lianetm commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkThreadTest.java:
```diff
@@ -72,68 +64,98 @@
+        when(networkClientDelegate.addAll(pollResult1)).thenReturn(pollResult1.timeUntilNextPollMs);
+        consumerNetworkThread.runOnce();
+
+        verify(networkClientDelegate).poll(exampleTime < 5001 ? exampleTime : 5000, time.milliseconds());
```
It took me a bit to understand this test, and the trick is that this 5000 is the MAX_POLL_TIMEOUT_MS, so would be clearer to use the constant I believe (here and on the param 5001 -> MAX_POLL_TIMEOUT_MS + 1).

If using the constant then here would probably be clearer to have `Math.min(exampleTime, MAX_POLL_TIMEOUT_MS)`

lianetm commented:
Regarding the PR description, I would suggest to update:

> Removed testEnsureMetadataUpdateOnPoll() since it was doing integration testing~~, could not get a unit test to work for that one~~

You did just great removing it, as it was testing something that it's not the responsibility of this component, and that's exactly why a real unit test is not doable ;)

brenden20 committed:
Implementing suggestions

brenden20 committed:
Implement suggestion

brenden20 commented:
@lianetm implemented your suggestions on the description and testConsumerNetworkThreadPollTimeComputations(). Let me know how it looks

lianetm commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkThreadTest.java:
```diff
@@ -72,68 +64,99 @@
-        consumerNetworkThread = new ConsumerNetworkThread(
-                testBuilder.logContext,
+    static final int DEFAULT_HEARTBEAT_INTERVAL_MS = 1000;
+    static final long MAX_POLL_TIMEOUT_MS = 5000;
```
let's better use the constant defined in ConsumerNetworkThread.MAX_POLL_TIMEOUT_MS so that we don't have to maintain this one (and avoid a test break if that one changes)

lianetm replied:
I mean to remove this and reference ConsumerNetworkThread.MAX_POLL_TIMEOUT_MS when needed (like it's already done here for some tests)

lianetm commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkThreadTest.java:
```diff
@@ -72,68 +64,99 @@
-        coordinatorRequestManager = testBuilder.coordinatorRequestManager.orElseThrow(IllegalStateException::new);
-        consumerNetworkThread = new ConsumerNetworkThread(
-                testBuilder.logContext,
+    static final int DEFAULT_HEARTBEAT_INTERVAL_MS = 1000;
```
this seems to be only used in `testMaximumTimeToWait` right? if so we better keep it as a local var

lianetm commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkThreadTest.java:
```diff
@@ -72,68 +64,99 @@
+    }
+
+    @ParameterizedTest
+    @ValueSource(longs = {100, 4999, 5001})
```
```suggestion
    @ValueSource(longs = {100, 4999, ConsumerNetworkThread.MAX_POLL_TIMEOUT_MS + 1})
```

lianetm commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkThreadTest.java:
```diff
@@ -72,68 +64,99 @@
+        when(networkClientDelegate.addAll(pollResult1)).thenReturn(pollResult1.timeUntilNextPollMs);
+        consumerNetworkThread.runOnce();
+
+        verify(networkClientDelegate).poll(Math.min(exampleTime, MAX_POLL_TIMEOUT_MS), time.milliseconds());
```
```suggestion
        verify(networkClientDelegate).poll(Math.min(exampleTime, ConsumerNetworkThread.MAX_POLL_TIMEOUT_MS), time.milliseconds());
```
Let's use the constant to show the intention aligned with the param, but reusing the original one to avoid maintaining a copy here that would make the test break if the original changes.

brenden20 committed:
Updated testConsumerNetworkThreadPollTimeComputations()

Updates testConsumerNetworkThreadPollTimeComputations() and fixed a checkstyle error

brenden20 commented:
@lianetm I implemented your new suggestions, let me know if it looks good

lianetm commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkThreadTest.java:
```diff
@@ -211,178 +219,52 @@ public void testResetPositionsProcessFailureIsIgnored() {
-            }
-            return true;
-        };
+        assertFalse(networkClientDelegate.hasAnyPendingRequests());
```
I'm afraid we lost some changes here? we were verifying the networkClientDelegate.poll 2 times (and wanted to remove the assert on the mock)

brenden20 replied:
Whoops, yeah I highlighted the wrong lines by accident. Fixed that now

brenden20 committed:
Fixed error

lianetm reviewed (approved):
Thanks for changes @brenden20 , this LGTM.

Really nice how you got rid of the heavy use of unneeded spy objects that the ConsumerTestBuilder was bringing in. Thanks!

mjsax commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkThreadTest.java:
```diff
@@ -72,68 +64,97 @@
+    }
+
+    @Test
+    public void testEnsureCloseStopsRunningThread() {
```
Seems this is one of the newly added tests?

brenden20 replied:
Yes

mjsax commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkThreadTest.java:
```diff
@@ -72,68 +64,97 @@
+
+    @ParameterizedTest
+    @ValueSource(longs = {ConsumerNetworkThread.MAX_POLL_TIMEOUT_MS - 1, ConsumerNetworkThread.MAX_POLL_TIMEOUT_MS, ConsumerNetworkThread.MAX_POLL_TIMEOUT_MS + 1})
+    public void testConsumerNetworkThreadPollTimeComputations(long exampleTime) {
```
Seems this is the second newly added tests?

brenden20 replied:
Yes

mjsax commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkThreadTest.java:
```diff
@@ -72,68 +64,97 @@
+    public void testConsumerNetworkThreadPollTimeComputations(long exampleTime) {
+        List<Optional<? extends RequestManager>> list = new ArrayList<>();
+        list.add(Optional.of(coordinatorRequestManager));
+        list.add(Optional.of(heartbeatRequestManager));
```
Why do we only add two of the three request managers to this list?

brenden20 replied:
Really just to declutter, I thought it might make that specific test kind of long if I added every request manager. 2 was the minimum amount necessary to test the functionality of the test. It just needs to return the shortest timeUntilNextPollMs. So in order to test this I just did 2 request managers with different poll times to make sure it was being calculated correctly. 

I can add the offsetsRequestManager if needed.

lianetm replied:
It wouldn't hurt to have all managers, but agree that having just a subset > 1 should be enough, because the logic is not specific to the managers (the consumerNetworkThread is blind on the managers, it only knows it has a set of them wrapped in a `RequestManagers` collection)

mjsax commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkThreadTest.java:
```diff
@@ -148,35 +169,32 @@ public void testStartupAndTearDown() throws InterruptedException {
     }
 
     @Test
-    public void testApplicationEvent() {
```
It seems this test is rewritten as `testApplicationEventIsProcessed` further below? And `testRequestsTransferFromManagersToClientOnThreadRun` is another newly added test?

brenden20 replied:
Yes, this test as well as many of the other application event tests were rewritten into a single parameterized test. It gets the parameters from ```applicationEvents()```, where it supplies a stream of different kinds of events. This was done to condense the code a bit.

And ```testRequestsTransferFromManagersToClientOnThreadRun()``` is a new test

mjsax commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkThreadTest.java:
```diff
@@ -148,35 +169,32 @@ public void testStartupAndTearDown() throws InterruptedException {
     }
 
-    @Test
-    public void testMetadataUpdateEvent() {
```
Seems this test was effectively removed? Why?

brenden20 replied:
Refer to previous comment for all event tests that were removed

mjsax commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkThreadTest.java:
```diff
@@ -148,35 +169,32 @@ public void testStartupAndTearDown() throws InterruptedException {
-    }
 
-    @Test
-    public void testAsyncCommitEvent() {
```
Seems this test was effectively removed? Why?

mjsax commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkThreadTest.java:
```diff
@@ -148,35 +169,32 @@ public void testStartupAndTearDown() throws InterruptedException {
+            verify(applicationEventReaper).add((CompletableEvent<?>) e);
 
-    @Test
-    public void testSyncCommitEvent() {
```
Seems this test was effectively removed? Why?

mjsax commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkThreadTest.java:
```diff
@@ -190,15 +208,6 @@ public void testListOffsetsEventIsProcessed(boolean requireTimestamp) {
     }
 
-    @Test
-    public void testResetPositionsEventIsProcessed() {
```
Seems this test was effectively removed? Why?

Same for more test below

mjsax reviewed (commented):
I am not very familiar with the control flow of the consumer, but it seems this PR is doing much more than an "in place" rewrite of the test. Would be good to understand better why so many test methods got removed.

philipnee commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkThreadTest.java:
```diff
@@ -211,178 +220,52 @@ public void testResetPositionsProcessFailureIsIgnored() {
+    private static Stream<Arguments> applicationEvents() {
+        Time time1 = new MockTime();
+        Map<TopicPartition, OffsetAndMetadata> offset = new HashMap<>();
+        final long currentTimeMs = time1.milliseconds();
```
can we just make up `currentTimeMs` value when constructing the event?

philipnee reviewed (approved):
hi @brenden20 - i left a comment, otherwise lgtm.

brenden20 committed:
Merge branch 'apache:trunk' into 16001

brenden20 committed:
Implemented suggestion

brenden20 committed:
Merge branch '16001' of https://github.com/brenden20/kafka into 16001

brenden20 committed:
Merge branch 'apache:trunk' into 16001

brenden20 committed:
Merge branch 'apache:trunk' into 16001

brenden20 committed:
Merge branch 'apache:trunk' into 16001

brenden20 committed:
Merge branch 'apache:trunk' into 16001

brenden20 commented:
@mjsax can you take a look?

brenden20 committed:
Merge branch 'apache:trunk' into 16001

mjsax commented:
Thanks for the PR @brenden20! Merged to `trunk`.

brenden20 commented:
@mjsax thank you for the review!
